File: backend/app/api/api_v1/endpoints/entity_values.py
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.schemas.entity_value import EntityValueCreate, EntityValue as EntityValueSchema
from app.services.entity_value import create_entity_value, get_latest_value
from app.api.deps import get_db

router = APIRouter()

@router.post("/", response_model=EntityValueSchema, status_code=status.HTTP_201_CREATED)
def create_entity_value_endpoint(entity_value_in: EntityValueCreate, db: Session = Depends(get_db)):
    entity_value = create_entity_value(
        db=db,
        entity_type=entity_value_in.entity_type,
        entity_id=entity_value_in.entity_id,
        scenario_id=entity_value_in.scenario_id,
        value=entity_value_in.value,
        recorded_at=entity_value_in.recorded_at or datetime.utcnow()
    )
    return entity_value

@router.get("/latest", response_model=EntityValueSchema)
def get_latest_entity_value(entity_type: str, entity_id: int, scenario_id: int, db: Session = Depends(get_db)):
    entity_value = get_latest_value(db, entity_type=entity_type, entity_id=entity_id, scenario_id=scenario_id)
    if not entity_value:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Entity value not found")
    return entity_value 

File: frontend/src/api/entityValues.ts
import { api } from './api';
import { EntityValue } from '@/types/finance';

export const entityValuesApi = {
  getEntityValues: async (scenarioId: number): Promise<EntityValue[]> => {
    const response = await api.get<{ data: EntityValue[] }>(`/entity-values?scenario_id=${scenarioId}`);
    return response.data;
  },

  getEntityValue: async (id: number): Promise<EntityValue> => {
    const response = await api.get<{ data: EntityValue }>(`/entity-values/${id}`);
    return response.data;
  },

  createEntityValue: async (data: {
    scenario_id: number;
    entity_type: string;
    entity_id: number;
    value: number;
    recorded_at?: string;
  }): Promise<EntityValue> => {
    const response = await api.post<{ data: EntityValue }>('/entity-values', data);
    return response.data;
  },

  updateEntityValue: async (id: number, data: {
    value: number;
    recorded_at?: string;
  }): Promise<EntityValue> => {
    const response = await api.put<{ data: EntityValue }>(`/entity-values/${id}`, data);
    return response.data;
  },

  deleteEntityValue: async (id: number): Promise<void> => {
    await api.delete(`/entity-values/${id}`);
  }
}; 